### PR TITLE
AB#393 : Marble unique secrets

### PR DESCRIPTION
**Pull request by Nirusu**
Can be reviewed together with the documentation at edgelesssys/marblerun.sh#6

### CODE REVIEW

File: coordinator/core/marbleapi.go [hunk 1]
```diff
@@ -107,8 +107,29 @@ func (c *Core) Activate(ctx context.Context, req *rpc.ActivationReq) (*rpc.Activ
 		SealKey:    Secret{Public: sealKey, Private: sealKey},
 	}
 
+	// Generate unique (= per marble) secrets
+	uniqueSecrets := make(map[string]Secret)
+	for k, v := range c.manifest.Secrets {
+		if v.Shared == false {
+			v.UUID = marbleUUID
```
Comment (thomasten) [inline coordinator/core/marbleapi.go:114]: As every secret gets the same uuid here, I'd propose to not have this as a field in the struct, but pass it as string to generateSecrets. Then you can also remove this filtering loop and the one in SetManifest and instead skip secrets in generateSecrets based on whether uuid is empty or not.

File: coordinator/core/marbleapi.go [hunk 2]
```diff
@@ -107,8 +107,29 @@ func (c *Core) Activate(ctx context.Context, req *rpc.ActivationReq) (*rpc.Activ
...
+	if err != nil {
+		c.zaplogger.Error("Could not generate specified secrets for the given manifest.", zap.Error(err))
+		return nil, err
+	}
+
+	// Union unique secrets with shared secrets
+	for k, v := range c.secrets {
+		if v.Shared == true {
+			secrets[k] = v
+		}
```
Comment (thomasten) [inline coordinator/core/marbleapi.go:128]:
```suggestion
		secrets[k] = v
```
always true

File: coordinator/core/marbleapi.go [hunk 3]
```diff
@@ -107,8 +107,29 @@ func (c *Core) Activate(ctx context.Context, req *rpc.ActivationReq) (*rpc.Activ
 		SealKey:    Secret{Public: sealKey, Private: sealKey},
```
Comment (thomasten) [inline coordinator/core/marbleapi.go:107]: Not directly related to this PR, but the function is a bit long. Can you refactor the code from
privk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
till
authSecrets := reservedSecrets{
into a new function?

File: coordinator/core/marbleapi_test.go [hunk 4]
```diff
@@ -32,6 +32,12 @@ import (
 	"google.golang.org/grpc/peer"
 )
 
+// Used to check if shared secrets are the same and non-shared secrets are indeed unique
+var backendFirstSharedCert x509.Certificate
```
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:36]: Pls move this into the marbleSpawner struct. Avoid global variables.

File: coordinator/core/marbleapi_test.go [hunk 5]
```diff
@@ -199,18 +210,80 @@ func (ms *marbleSpawner) newMarble(marbleType string, infraName string, shouldSu
...
+			backendFirstUniqueCert = *secretCertNonShared
+		}
+		ms.mutex.Unlock()
+
+		// Check if our certificate does actually expire 7 days, as specified, after it was generated
+		expectedNotBefore = secretCertNonShared.NotAfter.AddDate(0, 0, -7)
+		ms.assert.EqualValues(expectedNotBefore, secretCertNonShared.NotBefore)
+
+	} else if marbleType == "backend_other" {
+		// Validate generated shared secret certificate
```
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:256]: This block is duplicated 4 times. Try to refactor.

**Review by thomasten — CHANGES_REQUESTED** [review]

File: coordinator/core/core.go [hunk 6]
```diff
@@ -334,10 +334,25 @@ func (c *Core) generateSecrets(ctx context.Context, secrets map[string]Secret) (
...
+			// If a secret is shared, we generate a completely random key. If a secret is constrained to a marble, we derive a key from the core's private key.
+			if secret.Shared == true {
+				generatedValue = make([]byte, secret.Size/8)
+				_, err := rand.Read(generatedValue)
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				salt := secret.UUID.String() + name
+				secretKeyDerive, err := x509.MarshalPKCS8PrivateKey(c.privk)
```
[the remark this answers is not in the document]
Comment (Nirusu) [reply coordinator/core/core.go:348]: Nope, was just used to work on the marshalled versions from the previous PR that I immediately used it this way :D 

I'll change it.

File: coordinator/core/marbleapi_test.go [hunk 7]
```diff
@@ -295,6 +237,22 @@ func (ms *marbleSpawner) newMarbleAsync(marbleType string, infraName string, sho
 	}()
 }
 
+func (ms *marbleSpawner) verifyCertificateFromEnvironment(envName string, params *rpc.Parameters, opts x509.VerifyOptions) x509.Certificate {
```
Comment (Nirusu) [inline coordinator/core/marbleapi_test.go:240]: Shoud this return an error, panic or is the current way (assert inside the function) okay?

File: coordinator/core/core.go [hunk 8]
```diff
@@ -319,12 +320,20 @@ func (c *Core) getStatus(ctx context.Context) (int, string, error) {
...
 
 	// Generate secrets
 	for name, secret := range secrets {
+
+		// Skip secrets from wrong context
+		if secret.Shared && id != uuid.Nil {
+			continue
+		} else if !secret.Shared && id == uuid.Nil {
+			continue
+		}
```
Comment (thomasten) [inline coordinator/core/core.go:335]:
```suggestion
		if secret.Shared != (id == uuid.Nil) {
			continue
		}
```
I find this easier to read, but if you disagree, you can keep it as is.

File: coordinator/core/marbleapi_test.go [hunk 9]
```diff
@@ -222,6 +237,22 @@ func (ms *marbleSpawner) newMarbleAsync(marbleType string, infraName string, sho
...
+	p, _ := pem.Decode([]byte(params.Env[envName]))
+	ms.assert.NotNil(p)
+	certificate, err := x509.ParseCertificate(p.Bytes)
+	ms.assert.NotNil(p)
+	ms.assert.NoError(err)
```
Comment (thomasten) [inline coordinator/core/marbleapi_test.go:245]:
```suggestion
	ms.require.NotNil(p)
	certificate, err := x509.ParseCertificate(p.Bytes)
	ms.require.NoError(err)
```

**Review by thomasten — CHANGES_REQUESTED** [review]

File: coordinator/core/marbleapi_test.go (hunk 7 above)
Comment (thomasten) [reply coordinator/core/marbleapi_test.go:240]: okay, but see comment regarding `require` below

File: coordinator/core/core.go (hunk 8 above)
Comment (Nirusu) [reply coordinator/core/core.go:335]: That's fine, I'll take it

File: coordinator/core/marbleapi_test.go (hunk 9 above)
Comment (Nirusu) [reply coordinator/core/marbleapi_test.go]: I left the root CA check & NotBefore check as asserts, is this okay? Not quite sure when to use assert and when to use require...

Comment (thomasten) [reply coordinator/core/marbleapi_test.go]: Yep. require stops the test whereas assert just marks it as failed. So you use require if the rest of the test depends on the result. E.g., if p is nil, there would be a nil dereference.

**Review by thomasten — APPROVED** [review]

**Pull request merged**